varnish-sys/src/vcl/ws.rs

```diff
@@ -41,105 +45,129 @@ impl<'a> Workspace<'a> {
-        let ws = unsafe { validate_ws(self.raw) };
-        let al = align_of::<*const c_void>();
-        let aligned_sz = ((sz + al - 1) / al) * al;
+        #[cfg(not(test))]
+        {
+            ffi::WS_Alloc(validate_ws(self.raw), size as u32)
+        }
 
-        unsafe {
+        #[cfg(test)]
+        {
+            // `WS_Alloc` is a private part of `varnishd`, not the Varnish library,
+            // so it is only available if the output is a `cdylib`.
+            // When testing, VMOD is a lib or a bin,
+            // so we have to fake our own allocator.
+            let ws = validate_ws(self.raw);
+            let align = align_of::<*const c_void>();
+            let aligned_sz = ((size + align - 1) / align) * align;
             if ws.e.offset_from(ws.f) < aligned_sz as isize {
-                Err(VclError::String(format!(
-                    "not enough room for {aligned_sz} (rounded up from {sz}). f: {:?}, e: {:?}",
-                    ws.f, ws.e
-                )))
+                ptr::null_mut()
             } else {
-                let buf = from_raw_parts_mut(ws.f.cast::<u8>(), aligned_sz);
+                let p = ws.f.cast::<c_void>();
                 ws.f = ws.f.add(aligned_sz);
-                Ok(buf)
+                p
             }
         }
     }
 
-    /// Check if a byte slice is allocated in the workspace
-    pub fn is_slice_allocated(&self, slice: &[u8]) -> bool {
-        // FIXME: why check for slice size + 1?  In case of a string,
-        //  the caller should pass `CStr::to_bytes_with_nul`.
-        //  In case of a byte buffer, there shouldn't be an extra NULL byte
-        unsafe {
-            WS_Allocated(
-                self.raw,
-                slice.as_ptr().cast::<c_void>(),
-                slice.len() as isize + 1,
-            ) == 1
+    /// Check if a pointer is part of the current workspace
+    pub fn contains(&self, data: &[u8]) -> bool {
+        unsafe { WS_Allocated(self.raw, data.as_ptr().cast(), data.len() as isize) == 1 }
+    }
+
+    /// Allocate `[u8; size]` array on Workspace.
+    /// Returns a reference to uninitialized buffer, or an out of memory error.
+    ///
+    /// Will panic if `size` is 0.
+    pub fn alloc_buffer(&mut self, size: usize) -> Result<&'a mut [MaybeUninit<u8>], VclError> {
+        let ptr = unsafe { self.alloc(size) };
+        if ptr.is_null() {
+            Err(VclError::WsOutOfMemory(size))
+        } else {
+            Ok(unsafe { from_raw_parts_mut(ptr.cast(), size) })
         }
     }
 
-    /// Copy any `AsRef<[u8]>` into the workspace
-    pub fn copy_bytes<T: AsRef<[u8]>>(&mut self, src: &T) -> Result<&'a [u8], VclError> {
-        let buf = src.as_ref();
-        let l = buf.len();
+    /// Allocate memory on Workspace, and move a value into it.
+    /// The value will be dropped in case of out of memory error.
+    pub fn alloc_value<T>(&mut self, value: T) -> Result<&'a mut T, VclError> {
```

Comment:
I think there's a problem here: the workspace will be wiped at the end of the task, and `rust` will never get the chance to run `drop()`, will it?

Reply:
You mean `T::drop()` ?  This is correct - but workspace does not have any callback cleanup mechanism, right? And if so, we essentially only want "un-cleanable" types to be copied into workspace.  If so, we may need to mark this as unsafe too.  Current usage of this method is to allocate space for `vrt_blob` and `vrt_backend_probe` 

P.S. should this method be called `copy_value`?

Reply:
I think I would get rid of it actually, and only allow to create `&[u8]`, anything else is unsafe, and possibly somewhat useless at the moment

Reply:
Agreed, I will make it private (i use it in a few places internally) and rename it to `copy_value`